File: auth/app/src/main/java/com/google/firebase/quickstart/auth/java/MultiFactorSignInActivity.java
package com.google.firebase.quickstart.auth.java;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.FirebaseException;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.MultiFactorInfo;
import com.google.firebase.auth.MultiFactorResolver;
import com.google.firebase.auth.PhoneAuthCredential;
import com.google.firebase.auth.PhoneAuthOptions;
import com.google.firebase.auth.PhoneAuthProvider;
import com.google.firebase.auth.PhoneMultiFactorGenerator;
import com.google.firebase.auth.PhoneMultiFactorInfo;
import com.google.firebase.quickstart.auth.R;
import com.google.firebase.quickstart.auth.databinding.ActivityMultiFactorSignInBinding;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import static android.text.TextUtils.isEmpty;

/**
 * Activity that handles MFA sign-in
 */
public class MultiFactorSignInActivity extends BaseActivity implements
        View.OnClickListener {

    private static final String KEY_VERIFICATION_ID = "key_verification_id";
    public static final String EXTRA_MFA_RESOLVER = "EXTRA_MFA_RESOLVER";

    private ActivityMultiFactorSignInBinding mBinding;

    private MultiFactorResolver mMultiFactorResolver;
    private PhoneAuthCredential mPhoneAuthCredential;
    private String mVerificationId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = ActivityMultiFactorSignInBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());

        if (savedInstanceState != null) {
            onRestoreInstanceState(savedInstanceState);
        }

        List<Button> phoneFactorButtonList = new ArrayList<>();
        phoneFactorButtonList.add(mBinding.phoneFactor1);
        phoneFactorButtonList.add(mBinding.phoneFactor2);
        phoneFactorButtonList.add(mBinding.phoneFactor3);
        phoneFactorButtonList.add(mBinding.phoneFactor4);
        phoneFactorButtonList.add(mBinding.phoneFactor5);

        for (Button button : phoneFactorButtonList) {
            button.setVisibility(View.GONE);
        }

        mBinding.finishMfaSignIn.setOnClickListener(this);

        mMultiFactorResolver = retrieveResolverFromIntent(getIntent());
        List<MultiFactorInfo> multiFactorInfoList = mMultiFactorResolver.getHints();

        for (int i = 0; i < multiFactorInfoList.size(); ++i) {
            PhoneMultiFactorInfo phoneMultiFactorInfo = (PhoneMultiFactorInfo) multiFactorInfoList.get(i);
            Button button = phoneFactorButtonList.get(i);
            button.setVisibility(View.VISIBLE);
            button.setText(phoneMultiFactorInfo.getPhoneNumber());
            button.setClickable(true);
            button.setOnClickListener(generateFactorOnClickListener(phoneMultiFactorInfo));
        }
    }

    @Override
    public void onSaveInstanceState(Bundle bundle) {
        super.onSaveInstanceState(bundle);
        bundle.putString(KEY_VERIFICATION_ID, mVerificationId);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        mVerificationId = savedInstanceState.getString(KEY_VERIFICATION_ID);
    }

    private View.OnClickListener generateFactorOnClickListener(PhoneMultiFactorInfo phoneMultiFactorInfo) {
        return new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                PhoneAuthProvider.verifyPhoneNumber(
                        PhoneAuthOptions.newBuilder()
                                .setActivity(MultiFactorSignInActivity.this)
                                .setMultiFactorSession(mMultiFactorResolver.getSession())
                                .setMultiFactorHint(phoneMultiFactorInfo)
                                .setCallbacks(generateCallbacks())
                                // A timeout of 0 disables SMS-auto-retrieval.
                                .setTimeout(0L, TimeUnit.SECONDS)
                                .build());
            }
        };
    }

    private PhoneAuthProvider.OnVerificationStateChangedCallbacks generateCallbacks() {
        return new PhoneAuthProvider.OnVerificationStateChangedCallbacks() {
            @Override
            public void onVerificationCompleted(@NonNull PhoneAuthCredential phoneAuthCredential) {
                MultiFactorSignInActivity.this.mPhoneAuthCredential = phoneAuthCredential;
                mBinding.finishMfaSignIn.performClick();
                Toast.makeText(
                        MultiFactorSignInActivity.this, "Verification complete!", Toast.LENGTH_SHORT)
                        .show();
            }

            @Override
            public void onCodeSent(@NonNull String verificationId, @NonNull PhoneAuthProvider.ForceResendingToken token) {
                MultiFactorSignInActivity.this.mVerificationId = verificationId;
                mBinding.finishMfaSignIn.setClickable(true);
            }

            @Override
            public void onVerificationFailed(@NonNull FirebaseException e) {
                Toast.makeText(
                        MultiFactorSignInActivity.this, "Error: " + e.getMessage(), Toast.LENGTH_SHORT)
                        .show();
            }
        };
    }

    private MultiFactorResolver retrieveResolverFromIntent(Intent intent) {
        return intent.getParcelableExtra(EXTRA_MFA_RESOLVER);
    }

    private void onClickFinishSignIn() {
        if (mPhoneAuthCredential == null) {
            if (isEmpty(mBinding.smsCode.getText().toString())) {
                Toast.makeText(
                        MultiFactorSignInActivity.this, "You need to enter an SMS code.", Toast.LENGTH_SHORT)
                        .show();
                return;
            }
            mPhoneAuthCredential =
                    PhoneAuthProvider.getCredential(
                            mVerificationId, mBinding.smsCode.getText().toString());
        }
        mMultiFactorResolver
                .resolveSignIn(PhoneMultiFactorGenerator.getAssertion(mPhoneAuthCredential))
                .addOnSuccessListener(
                        new OnSuccessListener<AuthResult>() {
                            @Override
                            public void onSuccess(AuthResult authResult) {
                                setResult(Activity.RESULT_OK);
                                finish();
                            }
                        })
                .addOnFailureListener(
                        new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception e) {
                                Toast.makeText(
                                        MultiFactorSignInActivity.this, "Error: " + e.getMessage(), Toast.LENGTH_SHORT)
                                        .show();
                            }
                        });
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.finishMfaSignIn) {
            onClickFinishSignIn();
        }
    }
}

File: auth/app/src/main/java/com/google/firebase/quickstart/auth/kotlin/MultiFactorSignInActivity.kt
package com.google.firebase.quickstart.auth.kotlin

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.Toast
import com.google.firebase.FirebaseException
import com.google.firebase.auth.MultiFactorResolver
import com.google.firebase.auth.PhoneAuthCredential
import com.google.firebase.auth.PhoneAuthOptions
import com.google.firebase.auth.PhoneAuthProvider
import com.google.firebase.auth.PhoneMultiFactorGenerator
import com.google.firebase.auth.PhoneMultiFactorInfo
import com.google.firebase.quickstart.auth.R
import com.google.firebase.quickstart.auth.databinding.ActivityMultiFactorSignInBinding
import com.google.firebase.quickstart.auth.java.BaseActivity
import java.util.concurrent.TimeUnit

/**
 * Activity that handles MFA sign-in
 */
class MultiFactorSignInActivity : BaseActivity(), View.OnClickListener {

    private lateinit var binding: ActivityMultiFactorSignInBinding
    private lateinit var multiFactorResolver: MultiFactorResolver
    private var lastPhoneAuthCredential: PhoneAuthCredential? = null
    private var lastVerificationId: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMultiFactorSignInBinding.inflate(layoutInflater)
        setContentView(binding.root)
        savedInstanceState?.let { onRestoreInstanceState(it) }

        // Users are currently limited to having 5 second factors
        val phoneFactorButtonList = listOf(
                binding.phoneFactor1, binding.phoneFactor2, binding.phoneFactor3,
                binding.phoneFactor4, binding.phoneFactor5)
        for (button in phoneFactorButtonList) {
            button.visibility = View.GONE
        }

        binding.finishMfaSignIn.setOnClickListener(this)
        multiFactorResolver = retrieveResolverFromIntent(intent)

        val multiFactorInfoList = multiFactorResolver.hints
        for (i in multiFactorInfoList.indices) {
            val phoneMultiFactorInfo = multiFactorInfoList[i] as PhoneMultiFactorInfo
            val button = phoneFactorButtonList[i]
            button.visibility = View.VISIBLE
            button.text = phoneMultiFactorInfo.phoneNumber
            button.isClickable = true
            button.setOnClickListener(generateFactorOnClickListener(phoneMultiFactorInfo))
        }
    }

    public override fun onSaveInstanceState(bundle: Bundle) {
        super.onSaveInstanceState(bundle)
        bundle.putString(KEY_VERIFICATION_ID, lastVerificationId)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        lastVerificationId = savedInstanceState.getString(KEY_VERIFICATION_ID)
    }

    private fun generateFactorOnClickListener(phoneMultiFactorInfo: PhoneMultiFactorInfo): View.OnClickListener {
        return View.OnClickListener {
            PhoneAuthProvider.verifyPhoneNumber(
                    PhoneAuthOptions.newBuilder()
                            .setActivity(this@MultiFactorSignInActivity)
                            .setMultiFactorSession(multiFactorResolver.session)
                            .setMultiFactorHint(phoneMultiFactorInfo)
                            .setCallbacks(generateCallbacks()) // A timeout of 0 disables SMS-auto-retrieval.
                            .setTimeout(0L, TimeUnit.SECONDS)
                            .build())
        }
    }

    private fun generateCallbacks(): PhoneAuthProvider.OnVerificationStateChangedCallbacks {
        return object : PhoneAuthProvider.OnVerificationStateChangedCallbacks() {
            override fun onVerificationCompleted(phoneAuthCredential: PhoneAuthCredential) {
                lastPhoneAuthCredential = phoneAuthCredential
                binding.finishMfaSignIn.performClick()
                Toast.makeText(
                        this@MultiFactorSignInActivity, "Verification complete!", Toast.LENGTH_SHORT)
                        .show()
            }

            override fun onCodeSent(verificationId: String, token: PhoneAuthProvider.ForceResendingToken) {
                lastVerificationId = verificationId
                binding.finishMfaSignIn.isClickable = true
            }

            override fun onVerificationFailed(e: FirebaseException) {
                Toast.makeText(
                        this@MultiFactorSignInActivity, "Error: " + e.message, Toast.LENGTH_SHORT)
                        .show()
            }
        }
    }

    private fun retrieveResolverFromIntent(intent: Intent): MultiFactorResolver {
        return intent.getParcelableExtra(EXTRA_MFA_RESOLVER)!!
    }

    private fun onClickFinishSignIn() {
        if (lastPhoneAuthCredential == null) {
            if (TextUtils.isEmpty(binding.smsCode.text.toString())) {
                Toast.makeText(
                        this@MultiFactorSignInActivity, "You need to enter an SMS code.", Toast.LENGTH_SHORT)
                        .show()
                return
            }
            lastPhoneAuthCredential = PhoneAuthProvider.getCredential(
                    lastVerificationId!!, binding.smsCode.text.toString())
        }
        multiFactorResolver
                .resolveSignIn(PhoneMultiFactorGenerator.getAssertion(lastPhoneAuthCredential!!))
                .addOnSuccessListener {
                    setResult(Activity.RESULT_OK)
                    finish()
                }
                .addOnFailureListener { e ->
                    Toast.makeText(
                            this@MultiFactorSignInActivity, "Error: " + e.message, Toast.LENGTH_SHORT)
                            .show()
                }
    }

    override fun onClick(v: View) {
        if (v.id == R.id.finishMfaSignIn) {
            onClickFinishSignIn()
        }
    }

    companion object {
        private const val KEY_VERIFICATION_ID = "key_verification_id"
        const val EXTRA_MFA_RESOLVER = "EXTRA_MFA_RESOLVER"
    }
}